SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/Storage.cs

```diff
@@ -85,14 +86,57 @@ public Task PutLog(string channelName, Log log)
             return AddTaskToQueue(() =>
             {
                 var logJsonString = LogSerializer.Serialize(log);
-                _storageAdapter.Insert(TableName,
-                    new[] { ColumnChannelName, ColumnLogName },
-                    new List<object[]> {
-                        new object[] {channelName, logJsonString}
-                    });
+                var maxSize = _storageAdapter.GetMaxStorageSize();
+                var logSize = Encoding.UTF8.GetBytes(logJsonString).Length;
+                if (maxSize == -1)
+                {
+                    throw new StorageException("Failed to store a log to the database.");
+                }
+                if (maxSize <= logSize)
+                {
+                    throw new StorageException($"Log is too large ({logSize} bytes) to store in database. " +
+                            $"Current maximum database size is {maxSize} bytes.");
+                }
+                while (true)
+                {
+                    try
+                    {
+                        _storageAdapter.Insert(TableName,
+                            new[] { ColumnChannelName, ColumnLogName },
+                            new List<object[]> {
+                                new object[] {channelName, logJsonString}
+                            });
+                        return;
+                    }
+                    catch (StorageFullException e)
+                    {
+                        FreeSpaceForNewLog(logSize, channelName);
+                    }
+                }
             });
         }
 
+        private void FreeSpaceForNewLog(long newLogSize, string channelName)
+        {
+            var deletedLogsSize = 0L;
+            var currentFreeSpace = _storageAdapter.GetMaxStorageSize() - _storageAdapter.GetCurrentStorageSize();
+            while (true)
+            {
+                var oldestLog = _storageAdapter.Select(TableName, ColumnChannelName, channelName, string.Empty, null, 1, new string[] { ColumnIdName });
+                if (oldestLog != null && oldestLog.Count > 0)
+                {
+                    var logData = oldestLog[0];
+                    var logSize = Encoding.UTF8.GetBytes((string)logData[2]).Length;
+                    _storageAdapter.Delete(TableName, ColumnIdName, logData[0]);
+                    deletedLogsSize += logSize;
+                }
+                if (currentFreeSpace + deletedLogsSize > newLogSize)
```

Comment:
This is not quite correct. `newLogSize` is a size only for data in one column. It can be used only as estimate. Do other platforms have the same?